Setoid instances

jkopanski opened the pull request:
Following discussion we had last week I've started to define those
instances and it occurred to me that they would make nice addition to
the base library.

I've removed requirement for `Products (Set q)` instance from lawful
`CartesianClosed`.  It seems it wasn't used anywhere and I didn't know
what that would be here.

Provided instances quantify `Level`s individually as they're different
for different typeclasses.  Quantifiyng that at the module level would
mean that we wouldn't be able to substitute it differently.  For
example: `Products` can use `Setoid c ℓ` as `obj`.  But
`Exponentials` require that `obj` is `Setoid (c ⊔ ℓ) (c ⊔ ℓ)`, which
in turns requires `Products` for the same `obj`.  Fixing `c` and `ℓ`
at the module would prohibit `Exponentials` to get required
`Products`.

One module (`Function.Construct.Setoid`) was copied from
agda-categories, and there it was copied from old standard library, as
it got removed from the 2.0 version.

Last commit is my attempt at defining `Traced`, but I can't quite get it
work.  See the commit comment.

jkopanski committed:
Setoid functions instances

Following discussion we had last week I've started to define those
instances and it occurred to me that they would make nice addition to
the base library.

I've removed requirement for `Products (Set q)` instance from lawful
`CartesianClosed`.  It seems it wasn't used anywhere and I didn't know
what that would be here.

Provided instances quantify `Level`s individually as they're different
for different typeclasses.  Quantifiyng that at the module level would
mean that we wouldn't be able to substitute it differently.  For
example: `Products` can use `Setoid c ℓ` as `obj`.  But
`Exponentials` require that `obj` is `Setoid (c ⊔ ℓ) (c ⊔ ℓ)`, which
in turns requires `Products` for the same `obj`.  Fixing `c` and `ℓ`
at the module would prohibit `Exponentials` to get required
`Products`.

One module (`Function.Construct.Setoid`) was copied from
agda-categories, and there it was copied from old standard library, as
it got removed from the 2.0 version.

jkopanski committed:
Type check Function.Lift

Forgot to add this one previously.  It's not included in the
`Felix.Instances.Function` as the rest of the Function directory.

jkopanski committed:
Some Traced instance for setoid functions

By using categorical operations I don't have to come up with `cong`
proof manually.  It will be composed instead.
I don't know however if this definition is good one, so it'll be
commented out for the time being.

conal commented:
Thanks, Jakub! I think we will need this category.

conal commented:
> I've removed requirement for `Products (Set q)` instance from lawful
> `CartesianClosed`. It seems it wasn't used anywhere and I didn't know
> what that would be here.

Great. Perhaps this constraint was needed at one time and then no longer.

conal commented:
> Provided instances quantify `Level`s individually as they're different
> for different typeclasses. Quantifiyng that at the module level would
> mean that we wouldn't be able to substitute it differently. For
> example: `Products` can use `Setoid c ℓ` as `obj`. But
> `Exponentials` require that `obj` is `Setoid (c ⊔ ℓ) (c ⊔ ℓ)`, which
> in turns requires `Products` for the same `obj`. Fixing `c` and `ℓ`
> at the module would prohibit `Exponentials` to get required
> `Products`.

:D !

conal commented:
> One module (`Function.Construct.Setoid`) was copied from agda-categories, and there it was copied from old standard library, as it got removed from the 2.0 version.

Where was this functionality in the pre-2.0 agda-stdlib?

conal commented on src/Function/Construct/Setoid.agda:
```diff
@@ -0,0 +1,37 @@
+    where
+      module To = Setoid To
+
+  -- This doesn't really belong here, it should be in Function.Construct.Composition but that's in stdlib
+  -- so will need to be contributed to there first.
```
I don't understand this comment. Could you please rephrase?

jkopanski replied:
Oh, I took that file verbatim from [agda-categories](https://github.com/agda/agda-categories/blob/master/src/Function/Construct/Setoid.agda). What they mean is (I believe) that they wanted a nice infix operator for [Function.Construct.Composition.function](https://agda.github.io/agda-stdlib/v2.0/Function.Construct.Composition.html#5240) and therefore it should be in module `Function.Construct.Composition`, but that is in stdlib.

I think, that I didn't use this one, so perhaps we could just remove it?

conal replied:
Thanks for the clarification! I don't usually give the categorical operators any name other than the standard one (in the raw `Category` class).

conal commented on src/Felix/Instances/Setoid/Raw.agda:
```diff
@@ -0,0 +1,153 @@
+open import Data.Unit.Polymorphic using (tt) renaming (⊤ to ⊤′)
+open import Function using (Congruent; Func; _∘′_; flip′) renaming (_∘_ to _∘ᵈ_)
+open import Function.Construct.Composition as Comp
+open import Function.Construct.Constant as Const
```
Where is this module? I'm not finding it on my machine. Maybe I'm out of sync with the released 2.0.

jkopanski replied:
Yes, [it is in 2.0](https://agda.github.io/agda-stdlib/v2.0/Function.Construct.Constant.html), it's most likely that your one is out of date

jkopanski commented:
> > One module (`Function.Construct.Setoid`) was copied from agda-categories, and there it was copied from old standard library, as it got removed from the 2.0 version.
> 
> Where was this functionality in the pre-2.0 agda-stdlib?

I think it was [Function.Equality._⇨_](https://agda.github.io/agda-stdlib/v1.7.3/Function.Equality.html#2741).

jkopanski commented:
> > Provided instances quantify `Level`s individually as they're different
> > for different typeclasses. Quantifiyng that at the module level would
> > mean that we wouldn't be able to substitute it differently. For
> > example: `Products` can use `Setoid c ℓ` as `obj`. But
> > `Exponentials` require that `obj` is `Setoid (c ⊔ ℓ) (c ⊔ ℓ)`, which
> > in turns requires `Products` for the same `obj`. Fixing `c` and `ℓ`
> > at the module would prohibit `Exponentials` to get required
> > `Products`.
> 
> :D !

does that mean my explanation is incomprehensible? I don't know what you mean by that smiley

conal commented:
> > > One module (`Function.Construct.Setoid`) was copied from agda-categories, and there it was copied from old standard library, as it got removed from the 2.0 version.
> > 
> > 
> > Where was this functionality in the pre-2.0 agda-stdlib?
> 
> I think it was [Function.Equality._⇨_](https://agda.github.io/agda-stdlib/v1.7.3/Function.Equality.html#2741).

Thank you! That's the definition I was trying to remember. Any idea why it was removed?

jkopanski commented:
> > > > One module (`Function.Construct.Setoid`) was copied from agda-categories, and there it was copied from old standard library, as it got removed from the 2.0 version.
> > > 
> > > 
> > > Where was this functionality in the pre-2.0 agda-stdlib?
> > 
> > 
> > I think it was [Function.Equality._⇨_](https://agda.github.io/agda-stdlib/v1.7.3/Function.Equality.html#2741).
> 
> Thank you! That's the definition I was trying to remember. Do you have any idea why it was removed?

I was trying to look up the relevant [agda-categories commit history](https://github.com/agda/agda-categories/commit/5d04e4bdae88a824ad5c1a3bf07f97747903e458) in hopes to figure out more.  I know nothing for sure but my guess it was just forgotten.

Note that Jacques already opened this as an [stdlib PR](https://github.com/agda/agda-stdlib/pull/2240) with `v2.1` tag.

conal commented:
> > :D !
> 
> does that mean my explanation is incomprehensible? I don't know what you mean by that smiley

I meant I liked what you did.

conal commented:
There are a lot of opportunities to use standard categorical operations here. Did you consider doing so?

jkopanski commented:
> There are a lot of opportunities to use standard categorical operations here. Did you consider doing so?

Not really, it was mostly mechanistic translation from Function instances. I do however think that might be the case to cracking `Traced` as it would perhaps allow me to not reference directly the value that comes from well foundedness and instead do the same operation on the both sides of equality.

What did occur to me after this is that `Func.to` is homomorphism into Function, and that could be used to transfer properties

jkopanski committed:
Use more standard library

jkopanski committed:
Use more standard categorical operators

Actually I don't use that much _categorical operators_, as I couldn't
get proper categorical instances to work.  That is, Setoid categorical
laws are functions in the category of (Agda) function and types.
However I couldn't figure out the Level of Set that I should set to
`Felix.Instance.Function` in order to use those operators in laws
definitions.  So I settled for next best thing and used standard
library provided functions.  And oh boy, did I got surprised by the
results. Taking for example the `▵≈` law from the `Cartesian`. It
holds by `<_,_>` (which is the definition of _▵_ in category of
functions and types)!

This has completely stunned me. Law for the categorical operator holds
by the definition of this operator in different category!  I find it
amazing!

The same holds for `curry` in the `CartesianClosed`,  *but* I couldn't
figure it out for `∘≈` in the `Category`.  I wonder why?

I wonder if this has something to do with the fact that there is
trivial homomorphism from the category of `Setoid`s to category of
function and types via `⟦_⟧ = Func.to`?

In order to make `CartesianClosed` simplification to work I needed to
make argument to `Setoid` of setoid functions explicit.
Unfortunately in the upcoming PR it is implicit.

jkopanski commented:
> There are a lot of opportunities to use standard categorical operations here. Did you consider doing so?

I've tried this exercise and it blew my mind ;) see, comments attached to a related commit: 78e655a

conal commented:
> > There are a lot of opportunities to use standard categorical operations here. Did you consider doing so?
>
> I’ve tried this exercise and it blew my mind ;) see, comments attached to a related commit: [78e655a](https://github.com/conal/felix/commit/78e655aebf5c59089dbeaa2150371330a14ef791)

Terrific! It sounds like you’re having the experience I hoped you would. Start with poetic rephrasings, see what new hints appear, and follow where the beauty seems to lead.

conal commented:
> What did occur to me after this is that `Func.to` is homomorphism into Function, and that could be used to transfer properties

For properties to transfer, equivalence on the homomorphism’s source must be defined as equivalence of homomorphic images. When the homomorphism maps representations to their meanings, this definition of equivalence is (I think) a consequence of non-leaky abstraction. What notion or notions of morphism equivalence make sense in a **Setoid** category?

conal commented:
From commit [78e655a](https://github.com/conal/felix/commit/78e655aebf5c59089dbeaa2150371330a14ef791):

> This has completely stunned me. Law for the categorical operator holds by the definition of this operator in different category! I find it amazing!

I’m glad that you found this surprise yourself and that you enjoy the gift! I’ve stumbled onto such homomorphic gems often enough that I now know to dig for them. Finding them in logic/proofs is even more satisfying than in implementations, further rewarding the switch from Haskell to Agda.

jkopanski commented:
> > What did occur to me after this is that `Func.to` is homomorphism into Function, and that could be used to transfer properties
> 
> For properties to transfer, equivalence on the homomorphism’s source must be defined as equivalence of homomorphic images. When the homomorphism maps representations to their meanings, this definition of equivalence is necessary (I think) for non-leaky abstraction. What notion or notions of morphism equivalence make sense in a **Setoid** category?

IIUC that is exactly the case for equivalence in category of setoids: `∀ x → to f x ≈ to g x`. Ok now that I've pasted this in, it is not exactly the case, as in both categories we pointwise equality and it's not exact. So it should be rewritten in terms of `_≈_` from `Equivalent` for function instance?

conal commented:
> IIUC that is exactly the case for equivalence in category of setoids: `∀ x → to f x ≈ to g x`. Ok now that I've pasted this in, it is not exactly the case, as in both categories we pointwise equality and it's not exact. So it should be rewritten in terms of `_≈_` from `Equivalent` for function instance?

If we want equational properties (laws etc) to transfer backward through `to` in the standard homomorphic way, I think we’d have to define `_≈_ = _≈_ on to`, i.e., `f ≈ g = to f ≈ to g`, i.e., `f ≈ g = ∀ x → to f x ≡ to g x`. (You might enjoy reading through `Relation.Binary.Construct.On` if you haven’t already.) This `_≈_` definition doesn’t, however, match the intent of *Setoid*.

An alternative is to use a homomorphism going the other way: specify the function category (“**Set**”) via a homomorphism to the setoid category rather than vice versa.

conal commented on src/Felix/Instances/Setoid/Type.agda:
```diff
@@ -0,0 +1,52 @@
+  open import Felix.Object using (Coproducts; Exponentials; Products)
+
+  instance
+    products : ∀ {c ℓ} → Products (Setoid c ℓ)
```
Minor suggestion: Make `c` and `ℓ` parameters of the module instead. Ditto for `products` and `exponentials`.

conal commented on src/Felix/Laws.agda:
```diff
@@ -274,7 +274,7 @@ record CartesianClosed {obj : Set o} ⦃ _ : Products obj ⦄
                        ⦃ _ : R.Category _⇨′_ ⦄
                        ⦃ _ : R.Cartesian _⇨′_ ⦄
-                       ⦃ _ : R.CartesianClosed _⇨′_ ⦄
+                       ⦃ raw : R.CartesianClosed _⇨′_ ⦄
```
Probably merits a more specific name, considering the two previous instance parameters.

conal commented on src/Felix/Instances/Setoid/Raw.agda:
```diff
@@ -0,0 +1,90 @@
+  cartesianClosed = record
+    { curry = λ {A} {B} f → record
+      { to = λ a → record
+        { to = λ b → to f (a , b)
+        ; cong = λ x≈y → cong f (Setoid.refl A , x≈y)
```
`to` and `cong` can be written in terms of `curry′`.

conal commented on src/Felix/Instances/Setoid/Type.agda:
```diff
@@ -0,0 +1,52 @@
+_⟶_ : Setoid a ℓ₁ → Setoid b ℓ₂ → Set (a ⊔ b ⊔ ℓ₁ ⊔ ℓ₂)
+_⟶_ = Func
+
+open Func public
```
I don't think I'd `open` publicly.

jkopanski replied:
It exposed just `to` and `cong`. I've limited to `cong` given that we have an operator form for `to`. Unless `open`ing `public`ly does something else that I am not aware of?

conal replied:
I think `public` here means that the all of the `Func` fields & methods get exported from `Felix.Instances.Setoid.Type` and thus from `Felix.Instances.Setoid`, including `to`, `cong`, `isCongruent`, `Eq₁`, and `Eq₂`.

jkopanski replied:
Oh, you're completely right, forgot about other declarations. I've limited the `cong` anyway

conal commented on src/Function/Construct/Setoid.agda:
```diff
@@ -0,0 +1,31 @@
+
+{-# OPTIONS --without-K --safe #-}
+
+-- was not ported from old function hierarchy
```
Redundant comment?

jkopanski committed:
Remove duplicated comment

conal commented on src/Felix/Instances/Setoid/Raw.agda:
```diff
@@ -0,0 +1,91 @@
+
+module setoid-raw-instances where instance
+
+  equivalent : ∀ {c ℓ} → Equivalent (c ⊔ ℓ) {obj = Setoid c ℓ} _⟶_
+  equivalent = record
+    { _≈_ = λ {From} {To} f g →
+      let open Setoid To
+      in ∀ x → f ⟨$⟩ x ≈ g ⟨$⟩ x
+    ; equiv = λ {From} {To} →
+      let open Setoid To
+       in record
+         { refl = λ _ → refl
+         ; sym = λ f≈g x → sym (f≈g x)
+         ; trans = λ f≈g g≈h x → trans (f≈g x) (g≈h x)
+         }
+    }
```
Looks a lot like `Function.Construct.Setoid`. Perhaps `Felix.Equiv` should provide a helper to assemble an `Equivalent` from a `Setoid`.

conal replied:
`Equiv` provides the reverse as `≈setoid`.

jkopanski committed:
Reuse function setoid for equivalent instance

conal commented on src/Felix/Instances/Setoid/Raw.agda:
```diff
@@ -0,0 +1,84 @@
+
+  equivalent : ∀ {c ℓ} → Equivalent (c ⊔ ℓ) {obj = Setoid c ℓ} _⟶_
+  equivalent = record
+    { _≈_ = λ {From} {To} → Setoid._≈_ (setoid From To)
+    ; equiv = λ {From} {To} → Setoid.isEquivalence (setoid From To)
+    }
```
Thanks. I was hoping for a general mapping from `Setoid` to `Equivalent`, but now I realize that such a mapping would be tricky, since the setoid carrier might not have the required shape.

jkopanski replied:
I did try that for a bit, but couldn't figure it out

jkopanski committed:
Setoids distributive instance

conal commented on src/Felix/Instances/Setoid/Raw.agda:
```diff
@@ -0,0 +1,106 @@
+      ⦃ products {c} {c ⊔ ℓ} ⦄ ⦃ coproducts {c} {ℓ} ⦄
+      _⟶_
+      ⦃ category {c} {c ⊔ ℓ} ⦄ ⦃ cartesian {c} {c ⊔ ℓ} ⦄ ⦃ cocartesian {c} {ℓ} ⦄
+  distributive = record
+    { distribˡ⁻¹ = record
+      { to = λ (a , bc) → [ inj₁ ∘ᵈ (a ,_) , inj₂ ∘ᵈ (a ,_) ] bc
+      ; cong = λ where
+        (a₁≈a₂ , inj₁ b₁≈b₂) → inj₁ (a₁≈a₂ , b₁≈b₂)
+        (a₁≈a₂ , inj₂ c₁≈c₂) → inj₂ (a₁≈a₂ , c₁≈c₂)
+      }
+    ; distribʳ⁻¹ = record
+      { to =  λ (bc , a) → [ inj₁ ∘ᵈ (_, a) , inj₂ ∘ᵈ (_, a) ] bc
+      ; cong = λ where
+        (inj₁ b₁≈b₂ , a₁≈a₂) → inj₁ (b₁≈b₂ , a₁≈a₂)
+        (inj₂ c₁≈c₂ , a₁≈a₂) → inj₂ (c₁≈c₂ , a₁≈a₂)
+      }
+    }
```
How tidy could these definitions become if rewritten via the categorical vocabulary?

jkopanski replied:
oh, I took me way too long to notice, still can't replace `cong` one as I can't guide Agda to pick proper instances, which I myself don't know in full details. I was hoping agda would figure it out. All I see is the pattern for distribl/r

jkopanski replied:
Ok, the issue here is that constructors (`inj₁`/`inj₂`) used at `cong` are coming from [`Pointwise`](https://agda.github.io/agda-stdlib/v2.0/Data.Sum.Relation.Binary.Pointwise.html#998) and we don't have any category instance that would use this as coproducts*. I've tried using `map` from that module but I have trouble convincing Agda is complaining about stuff that I don't know how to solve.

* What would that category be? Category of [`REL A C`](https://agda.github.io/agda-stdlib/v2.0/Relation.Binary.Core.html#780)? [This one?](https://agda.github.io/agda-categories/Categories.Category.Instance.Rels.html)

conal replied:
> Ok, the issue here is that constructors (`inj₁`/`inj₂`) used at `cong` are coming from [`Pointwise`](https://agda.github.io/agda-stdlib/v2.0/Data.Sum.Relation.Binary.Pointwise.html#998) [...]

I don't know, and I like this question a lot! If we come up with a good answer, this whole Setoid category might come out beautifully.

conal replied:
Note that categorical products and coproducts in `Rels` are necessarily (isomorphic to) *sum* types.

conal commented on src/Felix/Instances/Setoid/Laws.agda:
```diff
@@ -0,0 +1,77 @@
+      ⦃ products {c} {c ⊔ ℓ} ⦄ ⦃ coproducts {c} {ℓ} ⦄
+      _⟶_
+      ⦃ raw = setoid-raw-instances.distributive {c} {ℓ} ⦄
+  distributive = record
+    { distribˡ∘distribˡ⁻¹ = λ where
+      {A} {B} {C} (_ , inj₁ _) → Setoid.refl (A ×ₛ (B ⊎ₛ C))
+      {A} {B} {C} (_ , inj₂ _) → Setoid.refl (A ×ₛ (B ⊎ₛ C))
+    ; distribˡ⁻¹∘distribˡ = λ where
+      {A} {B} {C} (inj₁ _) → Setoid.refl ((A ×ₛ B) ⊎ₛ (A ×ₛ C))
+      {A} {B} {C} (inj₂ _) → Setoid.refl ((A ×ₛ B) ⊎ₛ (A ×ₛ C))
+    ; distribʳ∘distribʳ⁻¹ = λ where
+      {A} {B} {C} (inj₁ _ , _) → Setoid.refl ((B ⊎ₛ C) ×ₛ A)
+      {A} {B} {C} (inj₂ _ , _) → Setoid.refl ((B ⊎ₛ C) ×ₛ A)
+    ; distribʳ⁻¹∘distribʳ = λ where
+      {A} {B} {C} (inj₁ _) → Setoid.refl ((B ×ₛ A) ⊎ₛ (C ×ₛ A))
+      {A} {B} {C} (inj₂ _) → Setoid.refl ((B ×ₛ A) ⊎ₛ (C ×ₛ A))
+    }
```
How pretty could these definitions look if expressed via categorical vocabulary?

jkopanski replied:
basically variation of the `(Setoid.refl ▿ Setoid.refl) ∘ exr` but the level inference for proper categorical instances are killing me

conal commented on src/Felix/Instances/Setoid/Raw.agda:
```diff
@@ -0,0 +1,106 @@
+      ⦃ products {c} {c ⊔ ℓ} ⦄ ⦃ coproducts {c} {ℓ} ⦄
+      _⟶_
+      ⦃ category {c} {c ⊔ ℓ} ⦄ ⦃ cartesian {c} {c ⊔ ℓ} ⦄ ⦃ cocartesian {c} {ℓ} ⦄
+  distributive = record
+    { distribˡ⁻¹ = record
+      { to = λ (a , bc) → [ inj₁ ∘ᵈ (a ,_) , inj₂ ∘ᵈ (a ,_) ] bc
+      ; cong = λ where
+        (a₁≈a₂ , inj₁ b₁≈b₂) → inj₁ (a₁≈a₂ , b₁≈b₂)
+        (a₁≈a₂ , inj₂ c₁≈c₂) → inj₂ (a₁≈a₂ , c₁≈c₂)
+      }
+    ; distribʳ⁻¹ = record
+      { to =  λ (bc , a) → [ inj₁ ∘ᵈ (_, a) , inj₂ ∘ᵈ (_, a) ] bc
+      ; cong = λ where
+        (inj₁ b₁≈b₂ , a₁≈a₂) → inj₁ (b₁≈b₂ , a₁≈a₂)
+        (inj₂ c₁≈c₂ , a₁≈a₂) → inj₂ (c₁≈c₂ , a₁≈a₂)
+      }
```
These definitions hint of tidier categorical formulations.

conal commented on src/Felix/Instances/Setoid/Type.agda:
```diff
@@ -0,0 +1,52 @@
+    coproducts = record
+      { ⊥ = record
+        { Carrier = ⊥′
+        ; _≈_ = λ _ _ → ⊤′
```
This one feels a little off to me. How about you?

conal commented:
@jkopanski I'm sorry for forgetting about this PR! I plan to go with #16, which was helpfully informed by your PR and will be easier to merge at this point Thanks again!

bsaul commented:
I borrowed some of what was done here in #16 (thanks @jkopanski). Maybe we close this PR?
